### ISSUE TITLE
Hundreds of  `error CA1416: This call site is reachable on all platforms. 'Xyz' is only supported on: 'Windows' 7.0 and later.` targeting net5.0-windows

### ISSUE THREAD

**Issue by RussKie** [dotnet/roslyn-analyzers#4983]
### Analyzer

**Diagnostic ID**: [CA1416](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/CA1416): `This call site is reachable on all platforms. 'xyz' is only supported on: 'Windows' 7.0 and later.`

### Analyzer source

**SDK**: 5.0.201

### Describe the bug

Hundreds of CA1416 despite targeting `net5.0-windows` (or `net5.0-windows7.0` as suggested in https://github.com/dotnet/roslyn-analyzers/issues/4699#issuecomment-804279890)

### Steps To Reproduce

Provide the steps to reproduce the behavior:
1. `git clone -v --recurse-submodules --depth 1 --no-single-branch --progress --branch Convert_to_.NET_5.0_build_CA1416 https://github.com/RussKie/gitextensions.git "<target forlder>"`
2. `dotnet build`
3. Observe hundreds of CA1416 (there are other NRT-related issues too)
    ![image](https://user-images.githubusercontent.com/4403806/112748929-d571a680-900a-11eb-8edf-487073ee8146.png)

### Expected behavior

No CA1416

## Additional context

I'm porting an app from net461 to net5.0.

**Comment by buyaa-n** [comment]
@RussKie this is not a bug, seems your repo is disabled AssemblyInfo.cs file generation (`<GenerateAssemblyInfo>false</GenerateAssemblyInfo>`) https://github.com/RussKie/gitextensions/blob/1ffc6f91f959c6e947963bc8e176f09de08375d4/Directory.Build.props#L15.

Which is the only way MSBuild could add the `SupportedOSPlatform("windows7.0")` attribute. Please read the [SDK expert's suggestion about this from here](https://github.com/dotnet/sdk/issues/14502#issuecomment-828805675). 

Now because it is disabled you need to add the attribute manually, seems your repo has a shared assembly info file (probably https://github.com/RussKie/gitextensions/blob/master/CommonAssemblyInfo.cs) please add [assembly: System.Runtime.Versioning.SupportedOSPlatform("windows7.0")] in that file or wherever it needs.

**Comment by RussKie** [comment]
Thank you for the link and thanks to @sygmond who provided an insight in https://github.com/dotnet/roslyn-analyzers/issues/4699#issuecomment-852453573. I'll give it a go once I have a spare moment. 

However I can't help but think that the current experience is far from desired, even @shanselman [stumbled on it](https://www.hanselman.com/blog/updating-floatingglucose-to-net-5-display-nightscout-or-dexcom-glucose-values-on-the-windows-desktop):
![image](https://user-images.githubusercontent.com/4403806/120432440-4479cb80-c3bd-11eb-9417-c376c0c6fd45.png)

The error message must point to docs that explain the issue in plain English and provides suggestions on how to fix it. We can't expect users to know that `GenerateAssemblyInfo=false` is unsupported, or expect them to go and search for some random reply somewhere on GitHub.

/cc: @dsplaisted

**Comment by dsplaisted** [comment]
@jeffhandley @mikadumont It does seem like we should do something better here.  Some ideas:

- If an assembly has no `SupportedOSPlatform` attributes, don't generate error messages for each call site.  Generate one (suppressible) message saying the assembly can't be analyzed for platform compatibility
- If `GenerateAssemblyInfo` is false and the platform compatibility analyzer is enabled, generate a single error message saying that the platform compatibility analyzer doesn't work when `GenerateAssemblyInfo` is false.

**Comment by jmarolf** [comment]
In addition, we could provide a codefix to just add this assembly attribute in response to the new diagnostic

**Comment by buyaa-n** [comment]
> The error message must point to docs that explain the issue in plain English and provides suggestions on how to fix it.

I will add this scenario in the analyzers doc

> If an assembly has no SupportedOSPlatform attributes, don't generate error messages for each call site. Generate one (suppressible) message saying the assembly can't be analyzed for platform compatibility

An assembly does not have to have an attribute, unattributed API/assembly is accounted as cross-platform, we should warn when cross-platform assembly accessing platform-specific API

> If GenerateAssemblyInfo is false and the platform compatibility analyzer is enabled, generate a single error message saying that the platform compatibility analyzer doesn't work when GenerateAssemblyInfo is false.

> In addition, we could provide a codefix to just add this assembly attribute in response to the new diagnostic

Probably it is better to be done by SDK analyzer because setting GenerateAssemblyInfo to false could also restrict other useful assembly attribute generation. But not sure if there are any other essential attributes generated by MSBuild. 

cc @terrajobst

**Comment by jmarolf** [comment]
> An assembly does not have to have an attribute, unattributed API/assembly is accounted as cross-platform, we should warn when cross-platform assembly accessing platform-specific API

Yes but the scenario (as I understand it) is this:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>WinExe</OutputType>
    <TargetFramework>net6.0-windows</TargetFramework>
	<GenerateAssemblyInfo>false</GenerateAssemblyInfo>
    <UseWPF>true</UseWPF>
  </PropertyGroup>

</Project>
```

The analyzers will now not operation correctly. The assembly _should_ be platform specific but it does not indicate that this is the case. We should have a separate analyzer that examines the TFM and emits a warning if the project targets a platform-specific TFM but does not have attributes that mark it as platform specific.

**Comment by RussKie** [comment]
IIRC `<TargetFramework>net6.0-windows</TargetFramework>` implies `'Windows' 7.0 and later`.

**Comment by RussKie** [comment]
![image](https://user-images.githubusercontent.com/4403806/120556278-a0853400-c43f-11eb-89a1-4f59a09a7b77.png)

**Comment by buyaa-n** [comment]
> We should have a separate analyzer that examines the TFM and emits a warning if the project targets a platform-specific TFM but does not have attributes that mark it as platform specific.

@jmarolf I am not opposing having such analyzer, my point is that analyzer should be on the SDK/MsBuild side. The CA1416 initially was evaluating TFM instead of relying on MsBuild, then because of limitation in roslyn-analyzers (i did not participate in the discussion about that, so do not remember clearly what was the limitation, @terrajobst might share more details), we decided to move that logic into MSBuild, there if we want to have an analyzer that should be MSBuild analyzer

**Comment by jeffhandley** [comment]
The behavior of the Platform Compatibility Analyzer is documented to capture this scenario now.
https://docs.microsoft.com/en-us/dotnet/standard/analyzers/platform-compat-analyzer#how-the-analyzer-recognizes-tfm-target-platforms

Turning off `<GenerateAssemblyInfo>` bypasses other features from the SDK as well. I wonder if we should treat this as more of an SDK/MSBuild issue, such that when that property is set to false, users are guided to understand the implications of what was just bypassed.  I see [this documentation](https://docs.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#generateassemblyinfo), but I wonder if there are other behaviors that need to be handled manually when this is disabled as well.

**Comment by terrajobst** [comment]
@dsplaisted 

> * If an assembly has no `SupportedOSPlatform` attributes, don't generate error messages for each call site.  Generate one (suppressible) message saying the assembly can't be analyzed for platform compatibility

That's not correct. Having no `SupportedOSPlatform` makes your project universally portable. You would get warnings when calling OS specific APIs without guarding them. That's the scenario for typical class libraries.

> * If `GenerateAssemblyInfo` is false and the platform compatibility analyzer is enabled, generate a single error message saying that the platform compatibility analyzer doesn't work when `GenerateAssemblyInfo` is false.

I don't think that's a good idea because it means people wouldn't be able to disable assembly info generation and use the analyzer, which they can today. The fix is to apply the necessary targeting attributes manually.

**Comment by RussKie** [comment]
> The behavior of the Platform Compatibility Analyzer is documented to capture this scenario now.
> [docs.microsoft.com/en-us/dotnet/standard/analyzers/platform-compat-analyzer#how-the-analyzer-recognizes-tfm-target-platforms](https://docs.microsoft.com/en-us/dotnet/standard/analyzers/platform-compat-analyzer#how-the-analyzer-recognizes-tfm-target-platforms)

👍 
The analyzer's error message needs to point to that doc.

**Comment by terrajobst** [comment]
> 👍
> The analyzer's error message needs to point to that doc.

It does. Click on the diagnostic ID. If you're thinking that this URL isn't discoverable, then [I think you're correct](https://github.com/dotnet/roslyn/issues/47310).

**Comment by buyaa-n** [comment]
Sounds like we can close this issue as the doc is updated and we are not adding/updating the analyzer as suggested

**Comment by SoftCircuits** [comment]
I'm still hitting this issue when I upgrade a WinForms project to .NET 8. If I set `GenerateAssemblyInfo` to true, then I get errors about duplicate definitions. I've found several pages talking about this, but it's not apparent what the correct fix is.

If Microsoft is going to make a breaking change, it really should be documented in a way that don't require me to stop my program and troubleshoot this change.

**Comment by buyaa-n** [comment]
> I'm still hitting this issue when I upgrade a WinForms project to .NET 8. If I set `GenerateAssemblyInfo` to true, then I get errors about duplicate definitions. I've found several pages talking about this, but it's not apparent what the correct fix is.

Not sure if you already found a solution, if not: when `GenerateAssemblyInfo` set to true `AssemblyInfo.cs` file is generated by MSBuild with the required assembly attributes like `[assembly: SupportedOSPlatform("windows")]`, you might have had a custom `AssemblyInfo.cs` file with some assembly attributes, therefore those attributes are duplicated. If that is the case instead of enabling `GenerateAssemblyInfo` add `[assembly: SupportedOSPlatform("windows")]` into your custom `AssemblyInfo.cs` file 

> If Microsoft is going to make a breaking change, it really should be documented in a way that doesn't require me to stop working on my code to troubleshoot this change. There should be a dead easy workaround.

Not sure if it was WinForms breaking change introduced in .NET 8.0 cc @JeremyKuhne @merriemcgaw

**Comment by SoftCircuits** [comment]
@buyaa-n In my case, the *custom* *AssemblyInfo.cs* file is in the *Properties* folder. And it seems to be working after I delete this file. Note that this isn't a file I created. It seems I have like a dozen WinForms apps that I'm updating to .NET 8.0, and they all have this file (and this same issue).

Anyway, maybe this will be helpful for someone else.
